### ISSUE TITLE
No Output on Run

### ISSUE THREAD

**Issue by Welding-Torch** [takinekotfs/jupyter-cpp-kernel#8]
**Describe the bug**
I installed the jupyter-cpp-kernel on Windows using pip, and when I launched jupyter it detected the C++ 14 kernel. I created a new file and typed in `std::cout << "Hello World" << std::endl;`. When I clicked 'Run', there was no output.

**To Reproduce**
1. Install using `pip install jupyter-cpp-kernel`
2. Launch Jupyter
3. New Notebook > C++ 14
4. Write Hello World in C++
5. Run and see **no output**

**Expected behavior**
To see `Hello World`

**Screenshots**
![image](https://github.com/takinekotfs/jupyter-cpp-kernel/assets/46340124/c1d1ac8e-0000-41ca-8b12-476a34d90dec)

**Desktop (please complete the following information):**
 - OS: Windows

**Additional context**
I have a hunch that this is a MinGW-w64 problem. My g++ exe is located at `C:\MinGW\bin\g++.exe`.

**Comment by takinekotfs** [comment]
You must follow these instructions, in order for me to investigate your issue

1. Try to run this command in `cmd`: `g++ -v`. Then you send the input there.
2. Send the Jupyter Notebook log (after you start Notebook, there's a command prompt window. It's the log). It is very helpful for us to investigate what really happened with your kernel.

**Comment by Welding-Torch** [comment]
Sorry for the late reply.

I've followed your instructions and copied the outputs into Pastebin. I use Pastebin for privacy reasons.

`g++ -v` output: https://pastebin.com/cHeZGMFJ
Jupyter Notebook Log: https://pastebin.com/hW5avp9S

While I do realise that you are using the Nixman builds of MinGW-w64 yourself, there shouldn't be an issue with my version. I'm using https://nuwen.net/mingw.html at the current moment. I would really like it if we could resolve this without me having to install another compiler.

**Comment by takinekotfs** [comment]
> Sorry for the late reply.
> 
> I've followed your instructions and copied the outputs into Pastebin. I use Pastebin for privacy reasons.
> 
> `g++ -v` output: https://pastebin.com/cHeZGMFJ Jupyter Notebook Log: https://pastebin.com/hW5avp9S
> 
> While I do realise that you are using the Nixman builds of MinGW-w64 yourself, there shouldn't be an issue with my version. I'm using https://nuwen.net/mingw.html at the current moment. I would really like it if we could resolve this without me having to install another compiler.

Your version of G++ 11.2.0 doesn't support `-ldl` (for linking additional compiling) for compiling - your Jupyter log already shows that.

```
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -ldl
```

Please consider using our [G++ version for Windows, version 13.2.0](https://github.com/takinekotfs/jupyter-cpp-kernel/releases/tag/gcc-13.2.0). It's fully supported for all features that the C++ kernel presented, and also supports new under-developing features `(We will release them in the near future)`.

**Comment by Welding-Torch** [comment]
I switched to Nixman builds directly (from [Nixman's](https://github.com/niXman/mingw-builds-binaries/releases/tag/13.2.0-rt_v11-rev0#:~:text=x86_64%2D13.2.0%2Drelease%2Dposix%2Dseh%2Ducrt%2Drt_v11%2Drev0.7z) ) as it's fully open-source, but I'm still getting the same 'no output on run' issue. 

Jupyter Log: https://pastebin.com/AywVFhS3

**Comment by takinekotfs** [comment]
Ok, I may investigate and remove `-ldl` in the near future. Hope you still can keep this conversation, I'll be back with the solved kernel.

**Comment by takinekotfs** [comment]
So far, I found your issue on Stack Overflow 😂

https://stackoverflow.com/questions/12455160/using-libdl-so-in-mingw

The reason that I keep no fix for your issue is, that the pre-compiled process `master.cpp` needs `libdl.so` to run properly. If I keep removing `-ldl`, the compiler will display this error

```shell
C:/mingw-gcc/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\takineko\AppData\Local\Temp\ccdj98F7.o:master.cpp:(.text+0xc67): undefined reference to `dlopen'
C:/mingw-gcc/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\takineko\AppData\Local\Temp\ccdj98F7.o:master.cpp:(.text+0xc77): undefined reference to `dlerror'
C:/mingw-gcc/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\takineko\AppData\Local\Temp\ccdj98F7.o:master.cpp:(.text+0xcb6): undefined reference to `dlerror'
C:/mingw-gcc/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\takineko\AppData\Local\Temp\ccdj98F7.o:master.cpp:(.text+0xcc9): undefined reference to `dlsym'
C:/mingw-gcc/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\takineko\AppData\Local\Temp\ccdj98F7.o:master.cpp:(.text+0xcd2): undefined reference to `dlerror'
```

You may need to install `libdl.so` into your process. Or, even better, using our [G++ version for Windows, version 13.2.0](https://github.com/takinekotfs/jupyter-cpp-kernel/releases/tag/gcc-13.2.0). It works out-of-box and is fully compatible with all features presented with the C++ kernel.

Hope you understand.

**Comment by takinekotfs** [comment]
If @Welding-Torch have no more comment or report-related issue, I recommend you close this issue. It's easier for me to track all current issues and know what need to do next.

**Comment by Welding-Torch** [comment]
Hello, I plan to include libdl.so in my configuration, but I'll need a couple of days to get to it due to some prior commitments.

I was wondering about the file size of your 64-bit installer for  [G++ version for Windows, version 13.2.0](https://github.com/takinekotfs/jupyter-cpp-kernel/releases/tag/gcc-13.2.0). It's 343 MB in size, whereas the latest [Nixman build](https://github.com/niXman/mingw-builds-binaries/releases/tag/13.2.0-rt_v11-rev0#:~:text=x86_64%2D13.2.0%2Drelease%2Dposix%2Dseh%2Ducrt%2Drt_v11%2Drev0.7z) for the same is only 69.6 MB. I'm curious if you could provide some insight into why there's such a significant difference in size?

Thank you for your assistance.

**Comment by takinekotfs** [comment]
> Hello, I plan to include libdl.so in my configuration, but I'll need a couple of days to get to it due to some prior commitments.
> 
> I was wondering about the file size of your 64-bit installer for [G++ version for Windows, version 13.2.0](https://github.com/takinekotfs/jupyter-cpp-kernel/releases/tag/gcc-13.2.0). It's 343 MB in size, whereas the latest [Nixman build](https://github.com/niXman/mingw-builds-binaries/releases/tag/13.2.0-rt_v11-rev0#:~:text=x86_64%2D13.2.0%2Drelease%2Dposix%2Dseh%2Ducrt%2Drt_v11%2Drev0.7z) for the same is only 69.6 MB. I'm curious if you could provide some insight into why there's such a significant difference in size?
> 
> Thank you for your assistance.

I don't really know how did I finish that (my brain is broken, tbh 🥲). 

The last thing I know is that the implementation of [all Linux libraries for MinGW](https://winlibs.com/). You can check it out.

**Comment by Welding-Torch** [comment]
Haha, its Winlibs! That's the same thing I use as a source for my C language installer at https://installc.org.

**Comment by takinekotfs** [comment]
Update issue progress: I'll remove GCC 13.2 which I have been releasing with that C++ kernel.

That doesn't mean the C++ kernel on Windows is retired, I just wanted to find a better way for all of you to use my kernel on as many platforms as I can.

**Comment by Welding-Torch** [comment]
> I just wanted to find a better way for all of you to use my kernel on as many platforms as I can.

That's a great initiative, I'm glad you're taking it. 🎉

**Comment by takinekotfs** [comment]
Hi @Welding-Torch

I recently just updated a document on how you can install GCC for Windows (it's a very detailed guide)

Check it out [here](https://github.com/takinekotfs/jupyter-cpp-kernel/blob/master/INSTALL_ON_WINDOWS.md).

**Comment by Welding-Torch** [comment]
Hey there,

I had a look at your guide. It's explanatory, though it may possibly be difficult to follow (depending on the skill level of the reader).

If I may, I just tried the C++ Kernel with my installer from https://installc.org/ . It works out-of-the-box, with 0 extra steps or configuration. It can be downloaded [here](https://github.com/Welding-Torch/installc/releases/latest/download/installc.exe) and it's source code is [here](https://github.com/Welding-Torch/installc/blob/main/Cinstaller1.iss). It uses `winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-16.0.6-mingw-w64ucrt-11.0.0-r1` for it's source . It's file size is 183 MB, which is even smaller than Winlibs's .zip file. 

Might I suggest that you recommend using Install C's installer for use with your project?

Thank you.

**Comment by takinekotfs** [comment]
> Hey there,
> 
> I had a look at your guide. It's explanatory, though it may possibly be difficult to follow (depending on the skill level of the reader).
> 
> If I may, I just tried the C++ Kernel with my installer from https://installc.org/ . It works out-of-the-box, with 0 extra steps or configuration. It can be downloaded [here](https://github.com/Welding-Torch/installc/releases/latest/download/installc.exe) and it's source code is [here](https://github.com/Welding-Torch/installc/blob/main/Cinstaller1.iss). It uses `winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-16.0.6-mingw-w64ucrt-11.0.0-r1` for it's source . It's file size is 183 MB, which is even smaller than Winlibs's .zip file.
> 
> Might I suggest that you recommend using Install C's installer for use with your project?
> 
> Thank you.

Great! If possible, you can promote your GCC/G++ in my Jupyter C++ kernel.

As I said, the C++ kernel is only fully compatible with Linux and macOS. I need someone to help me fully integrate this kernel in a Windows environment. I will look into your G++ installer (it may take a few days, depending on my schedule).

I just have a few questions before I can actually add your G++ installer to my project:

1. How can you maintain your installer? If there is a new release, how long will it take to release that compiler?
2. Are you planning to move to the `.msi` version, which integrates validated certificates?
3. What are your installer's limitations?

I asked the above questions just to understand both better.

I appreciate any help you can provide and thank you for taking the time to evaluate the C++ kernel and make it better.

**Comment by Welding-Torch** [comment]
Thanks for considering my installer!

To answer your questions:
1. I am planning to set up GitHub Actions or something similar to check Winlibs' GitHub for a new release, download it, and run the script to compile my installer with it automatically. Then increment version number and release it on Install C's GitHub Repo + Website. Until such a time as I implement this automation however, I will be updating the installer manually.
2. I had not considered that. I went for an .exe because it's simply a more common file extension and because my installation software defaulted to it. I will be looking into getting a certificate for my installer, though I'm told that may involve money.
3. Currently, only that it defaults to `C:\MinGW\` as it's install path. This was done so that I could call my installer 'one-click' and increase simplicity. In the future, I hope to allow users to select the directory in which they wish to install, while defaulting to `C:\Program Files\MinGW\` because that's neater than putting the folder in C:. 

Thank you for taking the time to evaluate the Install C installer.  I appreciate it.

**Comment by takinekotfs** [comment]
Hey @Welding-Torch, I just tried your GCC. And it was pretty good.

However, I have two significant recommendations for your installer:

1. You only have a GCC 64-bit installer. I recommend you to release the 32-bit version.
2. I recommend changing to the `.msi` installers because this extension makes the deployment on a large scale easier than using the `.exe`. You can skip this recommendation safely, it's just for enterprise features.

Thanks for taking the time in my project. Hope all of your work is better.

**Comment by Welding-Torch** [comment]
Thank you for the review!

I noticed that most of the PCs in use these days, including those around me, are 64-bit systems. As a result, I created a 64-bit installer. However, I will also make sure to include a 32-bit version in my internal task list for this project.

Will you endorse Install C as the recommended way to enable C support on Windows for use with jupyter-cpp-kernel?

I appreciate your input, and I hope your work is going smoothly as well.
